lab_solutions/lab03_3_rational_number_class.py

```diff
@@ -0,0 +1,166 @@
+#! /usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright © 2020 Attila Nagy <attila.nagy234@gmail.com>
+#
+# Distributed under terms of the MIT license.
+
+
+def gcd(x, y):
+    while x % y != 0:
+        prev_x = x
+        prev_y = y
+        x = prev_y
+        y = prev_x % prev_y
+    return y
+
+
+class RationalNumberValueError(ValueError):
+    pass
+
+
+class RationalNumber:
+    def __init__(self, p, q=1):
+        self.validate_numerator(p)
+        self.validate_denominator(q)
+        greatest_common_div = gcd(p, q)
+        if greatest_common_div != 1:
+            self._p = p // greatest_common_div
+            self._q = q // greatest_common_div
+        else:
+            self._p = p
+            self._q = q
+        if self._q < 0:
+            self.switch_signs()
+
+    def switch_signs(self):
+        self._p *= -1
+        self._q *= -1
+
+    def cross_multiple(self, other):
+        return self.p * other.q + self.q * other.p
+
+    def __add__(self, other):
+        numerator = self.cross_multiple(other)
+        denominator = self.q * other.q
+        return RationalNumber(numerator, denominator)
+
+    def __eq__(self, other):
+        return (self.p * other.q) == (self.q * other.p)
+
+    def __mul__(self, other):
+        if not isinstance(other, RationalNumber):
+            other = RationalNumber(other)
+        return RationalNumber(self.p * other.p, self.q * other.q)
+
+    def __imul__(self, other):
+        return self * other
+
+    def __iadd__(self, other):
+        return self + other
+
+    def __idiv__(self, other):
+        return self / other
+
+    def __invert__(self):
+        return RationalNumber(self.q, self.p)
+
+    def __truediv__(self, other):
+        return self * ~other
+
+    def __hash__(self):
+        return hash((self.p, self.q))
+
+    def __repr__(self):
+        return f'{self._p}/{self._q}'
+
+    def __abs__(self):
+        return abs(self.p / self.q)
+
+    @property
+    def p(self):
+        return self._p
+
+    @p.setter
+    def p(self, value):
+        self.validate_numerator(value)
+        self._p = value
```

Comment:
This and the setter of q does not normalize p and q. It should find their GCD like in the init function.